diff --git a/block/blk-merge.c b/block/blk-merge.c
--- a/block/blk-merge.c
+++ b/block/blk-merge.c
@@ -702,6 +702,9 @@ static struct request *attempt_merge(struct request_queue *q,
 	if (req->write_hint != next->write_hint)
 		return NULL;
 
+	if (req->ioprio != next->ioprio)
+		return NULL;
+
 	/*
 	 * If we are allowed to merge, then append bio list
 	 * from next to rq and release next. merge_requests_fn
@@ -863,6 +866,9 @@ bool blk_rq_merge_ok(struct request *rq, struct bio *bio)
 	if (rq->write_hint != bio->bi_write_hint)
 		return false;
 
+	if (rq->ioprio != bio_prio(bio))
+		return false;
+
 	/* Only merge if the crypt contexts are compatible */
 	if (!bio_crypt_ctx_compatible(bio, rq->bio))
 		return false;
*** End Patch
